File: clsPlotContainer.vb

Imports System.IO
Imports OxyPlot
Imports OxyPlot.Axes

Public Class clsPlotContainer

    Protected Enum ImageFileFormat
        PNG
        JPG
    End Enum

    Public Property AnnotationBottomLeft As String

    Public Property AnnotationBottomRight As String

    Public ReadOnly Property Plot As OxyPlot.PlotModel
        Get
            Return mPlot
        End Get
    End Property

    Public Property FontSizeBase As Integer

    Public ReadOnly Property SeriesCount As Integer
        Get
            If mPlot Is Nothing Then
                Return 0
            End If
            Return mPlot.Series.Count
        End Get
    End Property

    Public Property PlottingDeisotopedData As Boolean


    Protected mPlot As OxyPlot.PlotModel
    Protected mColorGradients As Dictionary(Of String, OxyPalette)

    ''' <summary>
    ''' Constructor
    ''' </summary>
    ''' <param name="thePlot"></param>
    ''' <remarks></remarks>
    Public Sub New(thePlot As OxyPlot.PlotModel)
        mPlot = thePlot
        FontSizeBase = 16
    End Sub

    ''' <summary>
    ''' Save the plot, along with any defined annotations, to a png file
    ''' </summary>
    ''' <param name="pngFilePath">Output file path</param>
    ''' <param name="width">PNG file width, in pixels</param>
    ''' <param name="height">PNG file height, in pixels</param>
    ''' <param name="resolution">Image resolution, in dots per inch</param>
    ''' <remarks></remarks>
    Public Sub SaveToPNG(pngFilePath As String, width As Integer, height As Integer, resolution As Integer)
        If String.IsNullOrWhiteSpace(pngFilePath) Then
            Throw New ArgumentOutOfRangeException(pngFilePath, "Filename cannot be empty")
        End If

        If Path.GetExtension(pngFilePath).ToLower() <> ".png" Then
            pngFilePath &= ".png"
        End If

        SaveToFileLoop(pngFilePath, ImageFileFormat.PNG, width, height, resolution)

    End Sub

    Public Sub SaveToJPG(jpgFilePath As String, width As Integer, height As Integer, resolution As Integer)
        If String.IsNullOrWhiteSpace(jpgFilePath) Then
            Throw New ArgumentOutOfRangeException(jpgFilePath, "Filename cannot be empty")
        End If

        If Path.GetExtension(jpgFilePath).ToLower() <> ".png" Then
            jpgFilePath &= ".jpg"
        End If

        SaveToFileLoop(jpgFilePath, ImageFileFormat.JPG, width, height, resolution)

    End Sub
    
    Protected Sub SaveToFileLoop(imageFilePath As String, fileFormat As ImageFileFormat, width As Integer, height As Integer, resolution As Integer)

        If mColorGradients Is Nothing OrElse mColorGradients.Count = 0 Then
            SaveToFile(imageFilePath, fileFormat, width, height, resolution)
            Return
        End If

        For Each colorGradient In mColorGradients
            Dim matchFound As Boolean

            For Each axis In mPlot.Axes
                Dim newAxis = TryCast(axis, LinearColorAxis)

                If newAxis Is Nothing Then
                    Continue For
                End If

                matchFound = True
                newAxis.Palette = colorGradient.Value
                newAxis.IsAxisVisible = True

                Dim fiBaseImageFile = New FileInfo(imageFilePath)

                Dim newFileName = Path.GetFileNameWithoutExtension(fiBaseImageFile.Name) & "_Gradient_" & colorGradient.Key & fiBaseImageFile.Extension
                newFileName = Path.Combine(fiBaseImageFile.DirectoryName, newFileName)

                SaveToFile(newFileName, fileFormat, width, height, resolution)
            Next

            If Not matchFound Then
                SaveToFile(imageFilePath, fileFormat, width, height, resolution)
                Return
            End If
        Next

    End Sub

    Protected Sub SaveToFile(imageFilePath As String, fileFormat As ImageFileFormat, width As Integer, height As Integer, resolution As Integer)

        Console.WriteLine("Saving " & Path.GetFileName(imageFilePath))

        ' Note that this operation can be slow if there are over 100,000 data points
        Dim plotBitmap = OxyPlot.Wpf.PngExporter.ExportToBitmap(mPlot, width, height, OxyPlot.OxyColors.White, resolution)

        Dim drawVisual = New DrawingVisual()
        Using drawContext = drawVisual.RenderOpen()

            Dim myCanvas = New Rect(0, 0, width, height)

            drawContext.DrawImage(plotBitmap, myCanvas)

            ' Add a frame
            Dim rectPen = New System.Windows.Media.Pen()
            rectPen.Brush = New SolidColorBrush(Colors.Black)
            rectPen.Thickness = 2

            drawContext.DrawRectangle(Nothing, rectPen, myCanvas)

            If Not String.IsNullOrWhiteSpace(AnnotationBottomLeft) Then
                AddText(AnnotationBottomLeft, drawContext, width, height, Windows.HorizontalAlignment.Left, Windows.VerticalAlignment.Bottom, 5)
            End If

            If Not String.IsNullOrWhiteSpace(AnnotationBottomRight) Then
                AddText(AnnotationBottomRight, drawContext, width, height, Windows.HorizontalAlignment.Right, Windows.VerticalAlignment.Bottom, 5)
            End If

            If PlottingDeisotopedData Then
                AddDeisotopedDataLegend(drawContext, width, height, 10, -30, 25)
            End If
        End Using

        Const DPI = 96

        Dim target = New RenderTargetBitmap(width, height, DPI, DPI, PixelFormats.Default)
        target.Render(drawVisual)

        Dim encoder As BitmapEncoder = Nothing

        Select Case fileFormat
            Case ImageFileFormat.PNG
                encoder = New PngBitmapEncoder()

            Case ImageFileFormat.JPG
                encoder = New JpegBitmapEncoder()
            Case Else
                Throw New ArgumentOutOfRangeException(fileFormat, "Unrecognized value: " + fileFormat.ToString())
        End Select

        If encoder IsNot Nothing Then
            encoder.Frames.Add(BitmapFrame.Create(target))

            Using outputStream = New FileStream(imageFilePath, FileMode.Create, FileAccess.Write)
                encoder.Save(outputStream)
            End Using
        End If

    End Sub

    Public Sub AddGradients(colorGradients As Dictionary(Of String, OxyPalette))
        mColorGradients = colorGradients
    End Sub

    Protected Sub AddDeisotopedDataLegend(
      drawContext As DrawingContext,
      canvasWidth As Integer,
      canvasHeight As Integer,
      offsetLeft As Integer,
      offsetTop As Integer,
      spacing As Integer)

        Const CHARGE_START = 1
        Const CHARGE_END = 6

        Dim usCulture = Globalization.CultureInfo.GetCultureInfo("en-us")
        ' Dim fontTypeface = New Typeface(New FontFamily("Arial"), FontStyles.Normal, System.Windows.FontWeights.Normal, FontStretches.Normal)
        Dim fontTypeface = New Typeface("Arial")

        Dim fontSizeEm = FontSizeBase + 3

        ' Write out the text 1+  2+  3+  4+  5+  6+  

        ' Create a box for the legend
        Dim rectPen = New System.Windows.Media.Pen()
        rectPen.Brush = New SolidColorBrush(Colors.Black)
        rectPen.Thickness = 1

        For chargeState = CHARGE_START To CHARGE_END

            Dim newBrush = New SolidColorBrush(GetColorByCharge(chargeState))

            Dim newText = New FormattedText(chargeState & "+", usCulture, FlowDirection.LeftToRight, fontTypeface, fontSizeEm, newBrush)

            Dim textRect = New Rect(0, 0, canvasWidth, canvasHeight)
            Dim position = textRect.Location

            position.X = mPlot.PlotArea.Left + offsetLeft + (chargeState - 1) * (newText.Width + spacing)
            position.Y = mPlot.PlotArea.Top + offsetTop

            If chargeState = CHARGE_START Then
                Dim legendBox = New Rect(position.X - 10, position.Y, CHARGE_END * (newText.Width + spacing) - spacing / 4, newText.Height)
                drawContext.DrawRectangle(Nothing, rectPen, legendBox)
            End If

            drawContext.DrawText(newText, position)
        Next

    End Sub

    Protected Sub AddText(
      textToAdd As String,
      drawContext As DrawingContext,
      canvasWidth As Integer,
      canvasHeight As Integer,
      hAlign As Windows.HorizontalAlignment,
      vAlign As Windows.VerticalAlignment,
      padding As Integer)

        Dim usCulture = Globalization.CultureInfo.GetCultureInfo("en-us")
        ' Dim fontTypeface = New Typeface(New FontFamily("Arial"), FontStyles.Normal, System.Windows.FontWeights.Normal, FontStretches.Normal)
        Dim fontTypeface = New Typeface("Arial")

        Dim fontSizeEm = FontSizeBase + 1

        Dim newText = New FormattedText(textToAdd, usCulture, FlowDirection.LeftToRight, fontTypeface, fontSizeEm, Brushes.Black)

        Dim textRect = New Rect(0, 0, canvasWidth, canvasHeight)
        Dim position = textRect.Location

        Select Case hAlign
            Case Windows.HorizontalAlignment.Left
                position.X += padding

            Case Windows.HorizontalAlignment.Center
                position.X += (textRect.Width - newText.Width) / 2

            Case Windows.HorizontalAlignment.Right
                position.X += textRect.Width - newText.Width - padding
        End Select

        Select Case vAlign
            Case Windows.VerticalAlignment.Top
                position.Y += padding

            Case Windows.VerticalAlignment.Center
                position.Y += (textRect.Height - newText.Height) / 2

            Case Windows.VerticalAlignment.Bottom
                position.Y += textRect.Height - newText.Height - padding
        End Select

        drawContext.DrawText(newText, position)
    End Sub

    Public Shared Function GetColorByCharge(charge As Integer) As Color
        Dim seriesColor As Color
        Select Case charge
            Case 1 : seriesColor = Colors.MediumBlue
            Case 2 : seriesColor = Colors.Red
            Case 3 : seriesColor = Colors.Green
            Case 4 : seriesColor = Colors.Magenta
            Case 5 : seriesColor = Colors.SaddleBrown
            Case 6 : seriesColor = Colors.Indigo
            Case 7 : seriesColor = Colors.LimeGreen
            Case 8 : seriesColor = Colors.CornflowerBlue
            Case Else : seriesColor = Colors.Gray
        End Select

        Return seriesColor

    End Function

    Protected Function PointSizeToEm(fontSizePoints As Integer) As Double
        Dim fontSizeEm = fontSizePoints / 12
        Return fontSizeEm
    End Function

    Protected Function PointSizeToPixels(fontSizePoints As Integer) As Integer
        Dim fontSizePixels = fontSizePoints * 1.33
        Return CInt(Math.Round(fontSizePixels, 0))
    End Function

End Class

File: clsPlotContainer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Windows;
using System.Windows.Media;
using System.Windows.Media.Imaging;
using OxyPlot;
using OxyPlot.Axes;

namespace MSFileInfoScanner
{
    using System.Globalization;

    /// <summary>
    /// Oxyplot container
    /// </summary>
    internal class clsPlotContainer : clsPlotContainerBase
    {
        private const double PIXELS_PER_DIP = 1.25;

        public const int DEFAULT_BASE_FONT_SIZE = 16;

        private enum ImageFileFormat
        {
            PNG,
            JPG
        }

        public OxyPlot.PlotModel Plot { get; }

        public int FontSizeBase { get; set; }

        public override int SeriesCount
        {
            get
            {
                if (Plot == null)
                {
                    return 0;
                }
                return Plot.Series.Count;
            }
        }

        private Dictionary<string, OxyPalette> mColorGradients;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="thePlot"></param>
        /// <param name="writeDebug"></param>
        /// <param name="dataSource"></param>
        /// <remarks></remarks>
        public clsPlotContainer(
            OxyPlot.PlotModel thePlot,
            bool writeDebug = false, string dataSource = "") : base(writeDebug, dataSource)
        {
            Plot = thePlot;
            FontSizeBase = DEFAULT_BASE_FONT_SIZE;
        }

        /// <summary>
        /// Save the plot, along with any defined annotations, to a png file
        /// </summary>
        /// <param name="pngFilePath">Output file path</param>
        /// <param name="width">PNG file width, in pixels</param>
        /// <param name="height">PNG file height, in pixels</param>
        /// <param name="resolution">Image resolution, in dots per inch</param>
        /// <remarks></remarks>
        public override void SaveToPNG(string pngFilePath, int width, int height, int resolution)
        {
            if (string.IsNullOrWhiteSpace(pngFilePath))
            {
                throw new ArgumentOutOfRangeException(pngFilePath, "Filename cannot be empty");
            }

            if (Path.GetExtension(pngFilePath).ToLower() != ".png")
            {
                pngFilePath += ".png";
            }

            SaveToFileLoop(pngFilePath, ImageFileFormat.PNG, width, height, resolution);

        }

        public void SaveToJPG(string jpgFilePath, int width, int height, int resolution)
        {
            if (string.IsNullOrWhiteSpace(jpgFilePath))
            {
                throw new ArgumentOutOfRangeException(jpgFilePath, "Filename cannot be empty");
            }

            if (Path.GetExtension(jpgFilePath).ToLower() != ".png")
            {
                jpgFilePath += "" +
                               ".jpg" +
                               "";
            }

            SaveToFileLoop(jpgFilePath, ImageFileFormat.JPG, width, height, resolution);

        }

        private void SaveToFileLoop(string imageFilePath, ImageFileFormat fileFormat, int width, int height, int resolution)
        {
            if (mColorGradients == null || mColorGradients.Count == 0)
            {
                SaveToFile(imageFilePath, fileFormat, width, height, resolution);
                return;
            }

            foreach (var colorGradient in mColorGradients)
            {
                var matchFound = false;

                foreach (var axis in Plot.Axes)
                {
                    var newAxis = axis as LinearColorAxis;

                    if (newAxis == null)
                    {
                        continue;
                    }

                    matchFound = true;
                    newAxis.Palette = colorGradient.Value;
                    newAxis.IsAxisVisible = true;

                    var fiBaseImageFile = new FileInfo(imageFilePath);

                    var newFileName = Path.GetFileNameWithoutExtension(fiBaseImageFile.Name) + "_Gradient_" + colorGradient.Key + fiBaseImageFile.Extension;
                    if (fiBaseImageFile.DirectoryName != null)
                    {
                        newFileName = Path.Combine(fiBaseImageFile.DirectoryName, newFileName);
                    }

                    SaveToFile(newFileName, fileFormat, width, height, resolution);
                }

                if (!matchFound)
                {
                    SaveToFile(imageFilePath, fileFormat, width, height, resolution);
                    return;
                }
            }

        }

        private void SaveToFile(string imageFilePath, ImageFileFormat fileFormat, int width, int height, int resolution)
        {
            if (string.IsNullOrWhiteSpace(imageFilePath))
                throw new ArgumentOutOfRangeException(nameof(imageFilePath), "imageFilePath cannot be empty");

            Console.WriteLine("Saving " + Path.GetFileName(imageFilePath));

            // Note that this operation can be slow if there are over 100,000 data points
            var plotBitmap = OxyPlot.Wpf.PngExporter.ExportToBitmap(Plot, width, height, OxyPlot.OxyColors.White, resolution);

            var drawVisual = new DrawingVisual();
            using (var drawContext = drawVisual.RenderOpen())
            {

                var myCanvas = new Rect(0, 0, width, height);

                drawContext.DrawImage(plotBitmap, myCanvas);

                // Add a frame
                var rectPen = new Pen
                {
                    Brush = new SolidColorBrush(Colors.Black),
                    Thickness = 2
                };

                drawContext.DrawRectangle(null, rectPen, myCanvas);

                if (!string.IsNullOrWhiteSpace(AnnotationBottomLeft)) {
                    AddText(AnnotationBottomLeft, drawContext, width, height, OxyPlot.HorizontalAlignment.Left, OxyPlot.VerticalAlignment.Bottom, 5);
                }

                if (!string.IsNullOrWhiteSpace(AnnotationBottomRight)) {
                    AddText(AnnotationBottomRight, drawContext, width, height, OxyPlot.HorizontalAlignment.Right, OxyPlot.VerticalAlignment.Bottom, 5);
                }

                if (PlottingDeisotopedData)
                {
                    AddDeisotopedDataLegend(drawContext, width, height, 10, -30, 25);
                }
            }

            const int DPI = 96;

            var target = new RenderTargetBitmap(width, height, DPI, DPI, PixelFormats.Default);
            target.Render(drawVisual);

            BitmapEncoder encoder;

            switch (fileFormat)
            {
                case ImageFileFormat.PNG:
                    encoder = new PngBitmapEncoder();

                    break;
                case ImageFileFormat.JPG:
                    encoder = new JpegBitmapEncoder();
                    break;
                default:
                    throw new ArgumentOutOfRangeException(fileFormat.ToString(), "Unrecognized value: " + fileFormat);
            }

            var bitMap = BitmapFrame.Create(target);

            encoder.Frames.Add(BitmapFrame.Create(target));

            using (var outputStream = new FileStream(imageFilePath, FileMode.Create, FileAccess.Write))
            {
                encoder.Save(outputStream);
            }

        }

        public void AddGradients(Dictionary<string, OxyPalette> colorGradients)
        {
            mColorGradients = colorGradients;
        }

        private void AddDeisotopedDataLegend(DrawingContext drawContext, int canvasWidth, int canvasHeight, int offsetLeft, int offsetTop, int spacing)
        {
            const int CHARGE_START = 1;
            const int CHARGE_END = 6;

            var usCulture = CultureInfo.GetCultureInfo("en-us");
            // var fontTypeface = new Typeface(new FontFamily("Arial"), FontStyles.Normal, System.Windows.FontWeights.Normal, FontStretches.Normal);
            var fontTypeface = new Typeface("Arial");

            var fontSizeEm = FontSizeBase + 3;

            // Write out the text 1+  2+  3+  4+  5+  6+

            // Create a box for the legend
            var rectPen = new Pen
            {
                Brush = new SolidColorBrush(Colors.Black),
                Thickness = 1
            };

            for (var chargeState = CHARGE_START; chargeState <= CHARGE_END; chargeState++)
            {
                var newBrush = new SolidColorBrush(GetColorByCharge(chargeState));

                var newText = new FormattedText(chargeState + "+", usCulture, FlowDirection.LeftToRight, fontTypeface, fontSizeEm, newBrush, null, PIXELS_PER_DIP);

                var textRect = new Rect(0, 0, canvasWidth, canvasHeight);
                var position = textRect.Location;

                position.X = Plot.PlotArea.Left + offsetLeft + (chargeState - 1) * (newText.Width + spacing);
                position.Y = Plot.PlotArea.Top + offsetTop;

                if (chargeState == CHARGE_START)
                {
                    var legendBox = new Rect(position.X - 10, position.Y, CHARGE_END * (newText.Width + spacing) - spacing / 4.0, newText.Height);
                    drawContext.DrawRectangle(null, rectPen, legendBox);
                }

                drawContext.DrawText(newText, position);
            }

        }

        private void AddText(string textToAdd, DrawingContext drawContext, int canvasWidth, int canvasHeight, OxyPlot.HorizontalAlignment hAlign, OxyPlot.VerticalAlignment vAlign, int padding)
        {
            var usCulture = CultureInfo.GetCultureInfo("en-us");
            // var fontTypeface = new Typeface(new FontFamily("Arial"), FontStyles.Normal, System.Windows.FontWeights.Normal, FontStretches.Normal);
            var fontTypeface = new Typeface("Arial");

            var fontSizeEm = FontSizeBase + 1;

            var newText = new FormattedText(textToAdd, usCulture, FlowDirection.LeftToRight, fontTypeface, fontSizeEm, Brushes.Black, null, PIXELS_PER_DIP);

            var textRect = new Rect(0, 0, canvasWidth, canvasHeight);
            var position = textRect.Location;

            switch (hAlign)
            {
                case OxyPlot.HorizontalAlignment.Left:
                    position.X += padding;

                    break;
                case OxyPlot.HorizontalAlignment.Center:
                    position.X += (textRect.Width - newText.Width) / 2;

                    break;
                case OxyPlot.HorizontalAlignment.Right:
                    position.X += textRect.Width - newText.Width - padding;
                    break;
            }

            switch (vAlign)
            {
                case OxyPlot.VerticalAlignment.Top:
                    position.Y += padding;

                    break;
                case OxyPlot.VerticalAlignment.Middle:
                    position.Y += (textRect.Height - newText.Height) / 2;

                    break;
                case OxyPlot.VerticalAlignment.Bottom:
                    position.Y += textRect.Height - newText.Height - padding;
                    break;
            }

            drawContext.DrawText(newText, position);
        }

        public static Color GetColorByCharge(int charge)
        {
            Color seriesColor;
            switch (charge)
            {
                case 1:
                    seriesColor = Colors.MediumBlue;
                    break;
                case 2:
                    seriesColor = Colors.Red;
                    break;
                case 3:
                    seriesColor = Colors.Green;
                    break;
                case 4:
                    seriesColor = Colors.Magenta;
                    break;
                case 5:
                    seriesColor = Colors.SaddleBrown;
                    break;
                case 6:
                    seriesColor = Colors.Indigo;
                    break;
                case 7:
                    seriesColor = Colors.LimeGreen;
                    break;
                case 8:
                    seriesColor = Colors.CornflowerBlue;
                    break;
                default:
                    seriesColor = Colors.Gray;
                    break;
            }

            return seriesColor;

        }

        private double PointSizeToEm(int fontSizePoints)
        {
            var fontSizeEm = fontSizePoints / 12;
            return fontSizeEm;
        }

        private int PointSizeToPixels(int fontSizePoints)
        {
            var fontSizePixels = fontSizePoints * 1.33;
            return (int)Math.Round(fontSizePixels, 0);
        }

    }
}

